tests/integration/sharding_tests/test_mongos.py

```diff
@@ -0,0 +1,154 @@
+async def test_build_and_deploy(ops_test: OpsTest, mongos_host_application_charm) -> None:
+    """Build and deploy a sharded cluster."""
+    mongodb_charm = await ops_test.build_charm(".")
+    await ops_test.model.deploy(
+        mongodb_charm,
+        num_units=1,
+        config={"role": "config-server"},
+        application_name=CONFIG_SERVER_APP_NAME,
+    )
+    await ops_test.model.deploy(
+        mongodb_charm, num_units=1, config={"role": "shard"}, application_name=SHARD_ONE_APP_NAME
+    )
+
+    await ops_test.model.deploy(
+        MONGOS_APP_NAME,
+        channel="6/edge",
+        revision=3,
+    )
+
+    # TODO: Future PR, once data integrator works with mongos charm deploy that charm instead of
+    # packing and deploying the charm in the application dir.
+    await ops_test.model.deploy(
+        mongos_host_application_charm, application_name=MONGOS_HOST_APP_NAME
+    )
+
+    await ops_test.model.wait_for_idle(
+        apps=[CONFIG_SERVER_APP_NAME, SHARD_ONE_APP_NAME, MONGOS_HOST_APP_NAME],
+        idle_period=20,
+        raise_on_blocked=False,  # cluster components are blocked waiting for integration.
+        timeout=TIMEOUT,
+        raise_on_error=False,
+    )
+
+
+@pytest.mark.abort_on_fail
+async def test_form_cluster(ops_test: OpsTest) -> None:
+    """Verifies that when the cluster is formed a new user is created."""
+    await ops_test.model.integrate(
+        f"{SHARD_ONE_APP_NAME}:{SHARD_REL_NAME}",
+        f"{CONFIG_SERVER_APP_NAME}:{CONFIG_SERVER_REL_NAME}",
+    )
+
+    await ops_test.model.integrate(
+        f"{MONGOS_APP_NAME}",
+        f"{MONGOS_HOST_APP_NAME}",
+    )
+
+    await ops_test.model.wait_for_idle(
+        apps=[MONGOS_HOST_APP_NAME, MONGOS_APP_NAME, SHARD_ONE_APP_NAME, CONFIG_SERVER_APP_NAME],
+        idle_period=20,
+        raise_on_blocked=False,
+        timeout=TIMEOUT,
+        raise_on_error=False,
+    )
+
+    mongos_client = await generate_mongodb_client(
+        ops_test, app_name=CONFIG_SERVER_APP_NAME, mongos=True
+    )
+    num_users = count_users(mongos_client)
+
+    await ops_test.model.integrate(
+        f"{MONGOS_APP_NAME}",
+        f"{CONFIG_SERVER_APP_NAME}",
+    )
+
+    await ops_test.model.wait_for_idle(
+        apps=[CONFIG_SERVER_APP_NAME, SHARD_ONE_APP_NAME, MONGOS_APP_NAME],
+        idle_period=20,
+        timeout=TIMEOUT,
+        raise_on_error=False,
+    )
+
+    num_users_after_integration = count_users(mongos_client)
+
+    assert (
+        num_users_after_integration > num_users
+    ), "Cluster did not create new users after integration."
+
+    (username, password) = await get_username_password(
+        ops_test, app_name=MONGOS_APP_NAME, relation_name=CLUSTER_REL_NAME
+    )
+    mongos_user_client = await generate_mongodb_client(
+        ops_test,
+        app_name=CONFIG_SERVER_APP_NAME,
+        mongos=True,
+        username=username,
+        password=password,
+    )
+
+    mongos_user_client.admin.command("dbStats")
+
+
```

Comment:
nit: (even though not useful in the last test, but just so it's not forgotten for when you add new tests after this one) 
```suggestion
@pytest.mark.abort_on_fail
```